src/core/device.c

```diff
@@ -114,12 +160,128 @@ static void device_done(Unit *u) {
         assert(d);
 
         device_unset_sysfs(d);
+        free(d->devname);
         d->wants_property = strv_free(d->wants_property);
 }
 
+static int root_cgroup_apply_block_device_io_cost_qos(Unit *root_slice_unit, Device *d) {
+        char buf[DECIMAL_STR_MAX(dev_t)*2+9+(6+DECIMAL_STR_MAX(uint64_t)+1)*10+(6+DECIMAL_STR_MAX(loadavg_t)+1)*2];
+        IOCostQos *io_cost_qos;
+        dev_t dev;
+        int r;
+
+        assert(root_slice_unit);
+        assert(d);
+
+        io_cost_qos = &d->io_cost_qos;
+
+        /* io_cost_qos enabled < 0 indicates that user did not set the enabled field */
+        if (io_cost_qos->enabled < 0)
+                return 0;
+
+        r = lookup_block_device(d->devname, &dev);
+        if (r < 0)
+                return r;
+
+        /* If ctrl is invalid but io_cost_qos is enabled there must be
+         * some misconfiguration by the user */
+        if (io_cost_qos->enabled && io_cost_qos->ctrl == _IO_COST_CTRL_INVALID)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Cannot set IOCostQOSEnabled=true without setting IOCostQOSCtrl=, parameters not valid");
```

Comment:
Could this check be made at parse-time? (Also see my comment about merging enable and ctrl together.)

Reply:
Not that I am aware of, if these are separate.

Reply:
I meant -- if this combo is invalid, it should not have been accepted as config in the first place (and not report it when applied to the root attribute).
But I'm more favorable of rethinking the provided directives to simply make invalid configs not expressible.

Reply:
I have reimplemented with only CTRL. The absence of CTRL will indicate disabled.